app/models/physical_server.rb

```diff
@@ -105,4 +107,8 @@ def v_availability
   def v_host_os
     host.try(:vmm_product).nil? ? N_("") : host.vmm_product
   end
+
+  def connected_switches
```

Comment:
So i see you're calling this [here](https://github.com/ManageIQ/manageiq-ui-classic/pull/3799/files#diff-4651fa89a3bd890efd1d5c3ec67fe397R11) but why not just call `:physical_switches` there?  I'm not sure why this would fix anything

Reply:
Not sure why, but I get the error that I said before:
```ruby
PhysicalServer Load (0.8ms)  SELECT "physical_servers".* FROM "physical_servers"
  PhysicalServer Inst Including Associations (587.0ms - 3rows)
  ComputerSystem Load (0.5ms)  SELECT "computer_systems".* FROM "computer_systems" WHERE "computer_systems"."managed_entity_type" = $1 AND "computer_systems"."managed_entity_id" IN (1, 2, 3)  [["managed_entity_type", "PhysicalServer"]]
  ComputerSystem Inst Including Associations (3.5ms - 3rows)
  Hardware Load (0.6ms)  SELECT "hardwares".* FROM "hardwares" WHERE "hardwares"."computer_system_id" IN (1, 2, 3)
  Hardware Inst Including Associations (16.4ms - 3rows)
  GuestDevice Load (0.4ms)  SELECT "guest_devices".* FROM "guest_devices" WHERE "switches"."type" IN ('PhysicalSwitch', 'ManageIQ::Providers::Lenovo::PhysicalInfraManager::PhysicalSwitch') AND "guest_devices"."hardware_id" IN (1, 2, 3)
ActiveRecord::StatementInvalid: PG::UndefinedTable: ERROR:  missing FROM-clause entry for table "switches"
LINE 1: ...LECT "guest_devices".* FROM "guest_devices" WHERE "switches"...
```
This error happens when you call `PhysicalServer.includes(:physical_switches)` which is what the topology service does.

With this method, I force the `#includes` method to call `physical_server.physical_switches` and then it works.

Basically, the `#includes` method tries to create separated queries for each join table, but `#physical_switches` does a single query joining every tables. This is why:
```ruby
physical_server_instance.physical_switches # works
PhysicalServer.includes(:physical_switches) # doesn't works
```

Reply:
@douglasgabriel looks like the assoc from physical_server needs to go through computer system, try this:

```
diff --git a/app/models/computer_system.rb b/app/models/computer_system.rb
index 062d9bb3cb..7f8ebc722d 100644
--- a/app/models/computer_system.rb
+++ b/app/models/computer_system.rb
@@ -3,4 +3,6 @@ class ComputerSystem < ApplicationRecord
 
   has_one :operating_system, :dependent => :destroy
   has_one :hardware, :dependent => :destroy
+
+  has_many :connected_physical_switches, :through => :hardware
 end
diff --git a/app/models/physical_server.rb b/app/models/physical_server.rb
index d309c35b31..ccbf8f7a3b 100644
--- a/app/models/physical_server.rb
+++ b/app/models/physical_server.rb
@@ -33,7 +33,7 @@ class PhysicalServer < ApplicationRecord
   virtual_column :v_availability, :type => :string, :uses => :host
   virtual_column :v_host_os, :type => :string, :uses => :host
 
-  has_many :physical_switches, :through => :hardware, :source => :connected_physical_switches
+  has_many :physical_switches, :through => :computer_system, :source => :connected_physical_switches
 
   def name_with_details
     details % {
@@ -107,8 +107,4 @@ class PhysicalServer < ApplicationRecord
   def v_host_os
     host.try(:vmm_product).nil? ? N_("") : host.vmm_product
   end
-
-  def connected_switches
-    physical_switches
-  end
 end
```

```
>> PhysicalServer.includes(:physical_switches)
PostgreSQLAdapter#log_after_checkout, connection_pool: size: 5, connections: 1, in use: 1, waiting_in_queue: 0
  PhysicalServer Load (0.4ms)  SELECT "physical_servers".* FROM "physical_servers"
  PhysicalServer Inst Including Associations (21.1ms - 1rows)
  SQL (1.7ms)  SELECT "computer_systems"."id" AS t0_r0, "computer_systems"."managed_entity_type" AS t0_r1, "computer_systems"."managed_entity_id" AS t0_r2, "computer_systems"."created_at" AS t0_r3, "computer_systems"."updated_at" AS t0_r4, "switches"."id" AS t1_r0, "switches"."name" AS t1_r1, "switches"."ports" AS t1_r2, "switches"."created_on" AS t1_r3, "switches"."updated_on" AS t1_r4, "switches"."uid_ems" AS t1_r5, "switches"."allow_promiscuous" AS t1_r6, "switches"."forged_transmits" AS t1_r7, "switches"."mac_changes" AS t1_r8, "switches"."switch_uuid" AS t1_r9, "switches"."shared" AS t1_r10, "switches"."mtu" AS t1_r11, "switches"."ems_id" AS t1_r12, "switches"."type" AS t1_r13, "switches"."health_state" AS t1_r14, "switches"."power_state" AS t1_r15 FROM "computer_systems" LEFT OUTER JOIN "hardwares" ON "hardwares"."computer_system_id" = "computer_systems"."id" LEFT OUTER JOIN "guest_devices" ON "guest_devices"."hardware_id" = "hardwares"."id" LEFT OUTER JOIN "physical_network_ports" "physical_network_ports_computer_systems_join" ON "physical_network_ports_computer_systems_join"."guest_device_id" = "guest_devices"."id" LEFT OUTER JOIN "physical_network_ports" ON "physical_network_ports"."connected_port_uid" = "physical_network_ports_computer_systems_join"."uid_ems" LEFT OUTER JOIN "switches" ON "switches"."id" = "physical_network_ports"."switch_id" AND "switches"."type" IN ('PhysicalSwitch', 'ManageIQ::Providers::Lenovo::PhysicalInfraManager::PhysicalSwitch') WHERE "computer_systems"."managed_entity_type" = $1 AND "switches"."type" IN ('PhysicalSwitch', 'ManageIQ::Providers::Lenovo::PhysicalInfraManager::PhysicalSwitch') AND "computer_systems"."managed_entity_id" = 1  [["managed_entity_type", "PhysicalServer"]]
  ComputerSystem Inst Including Associations (23.8ms - 1rows)
=> #<ActiveRecord::Relation [#<PhysicalServer id: 1, ems_id: nil, name: nil, type: nil, uid_ems: nil, ems_ref: nil, created_at: "2018-06-29 18:36:15", updated_at: "2018-06-29 18:36:15", health_state: nil, power_state: nil, hostname: nil, product_name: nil, manufacturer: nil, machine_type: nil, model: nil, serial_number: nil, field_replaceable_unit: nil, raw_power_state: nil, vendor: nil, location_led_state: nil, physical_rack_id: nil, ems_compliance_name: nil, ems_compliance_status: nil, physical_chassis_id: nil>]>
```

Reply:
Yeah, it works!